cli/src/main.rs

```diff
@@ -0,0 +1,761 @@
+    amounts[asset] -= amount;
+    assert!(amounts[asset] >= 0);
+    amounts[liab] += amount;
+    totals[asset] -= amount as u64;
+    // liabs weren't counted in totals!
+}
+
+fn move_amount(
+    amounts: &mut AccountTokenAmounts,
+    from: usize,
+    to: usize,
+    amount: i64,
+    totals: &mut [u64; 16],
+) {
+    assert!(from != to);
+    assert!(amount >= 0);
+    amounts[from] -= amount;
+    assert!(amounts[from] >= 0);
+    amounts[to] += amount;
+    totals[from] -= amount as u64;
+    totals[to] += amount as u64;
+}
+
+impl EquityFromSnapshot {
+    fn run(args: EquityFromSnapshotArgs) -> anyhow::Result<()> {
+        let late_changes = late_deposits_withdrawals(&args.late_changes)?;
+        let data = DataSource::new(args.sqlite.clone())?;
+
+        let group = data.load_group(args.group)?;
+        let mut cache = data.load_cache(group.mango_cache)?;
+
+        let ctx = EquityFromSnapshot { args, data, late_changes, group, cache };
+
+        let token_names: [&str; 16] = [
+            "MNGO", "BTC", "ETH", "SOL", "USDT", "SRM", "RAY", "COPE", "FTT", "ADA", "MSOL", "BNB",
+            "AVAX", "LUNA", "GMT", "USDC",
+        ];
+
+        println!("table,account,owner,{}", token_names.join(","));
+
+        let debug_print = |table: &str, data: &[AccountData]| {
+            for account in data.iter() {
+                println!(
+                    "{table},{},{},{}",
+                    account.mango_account,
+                    account.owner,
+                    account.amounts.iter().map(|v| v.to_string()).collect::<Vec<_>>().join(",")
+                );
+            }
+        };
+
+        let mut account_equities = ctx.snapshot_account_equities()?;
+
+        debug_print("snapshot", &account_equities);
+
+        ctx.apply_late_deposits_withdrawals(&mut account_equities)?;
+
+        debug_print("after dep/with", &account_equities);
+
+        ctx.skip_negative_equity_accounts(&mut account_equities)?;
+
+        let available_tokens: [bool; 15] = [
+            true, true, true, true, false, // usdt is gone
+            true, true, false, // cope delisted
+            true, false, // no spot ada
+            true, true, true, false, // luna delisted
+            true,
+        ];
+
+        // TODO: tentative numbers from "Repay bad Debt #2" proposal
+        let available_native_amounts: [u64; 15] = [
+            32409565000000,
+            281498000,
+            226000000,
+            761577910000000,
+            0,
+            2354260000000,
+            98295000000,
+            0,
+            11774000000,
+            0,
+            799155000000000,
+            60800000000,
+            180900000000,
+            0,
+            152843000000000,
+        ];
+
+        // Token prices at time of reimbursement
+        // Note that user equity at snapshot time is computed from the prices from the
+        // mango cache in the snapshot.
+        let reimbursement_prices: [I80F48; 16] = [
```

Comment:
Are these values going to be hardcoded right before the script gets kicked off? If so I presume multiple people would be verifying the values are correct? Would it be better to reference an oracle at runtime? 

I am a little bit scared that the script is relying on matching up indexes across multiple hardcoded arrays. They do seem to be correctly matched up though, and I suppose the reimbursement prices can have some margin of error.

Reply:
Yes, the prices would be frozen and updated as close to actual reimbursement as we can. I don't think pulling from an oracle is a good idea.

Constants: I agree and had meant to clean that up for a while. Done now.